contracts/bridge/contracts/BasicInboundChannel.sol

```diff
@@ -9,96 +9,87 @@ import "./interfaces/ILightClientBridge.sol";
         nonce = 0;
         lightClientBridge = _lightClientBridge;
     }
 
-    // TODO: Submit should take in all inputs required for verification,
     function submit(
         Message[] memory messages,
-        bytes memory beefyMMRLeaf,
+        BeefyMMRLeaf memory beefyMMRLeaf,
         uint256 beefyMMRLeafIndex,
         uint256 beefyMMRLeafCount,
         bytes32[] memory peaks,
         bytes32[] memory siblings 
     ) public {
+        bytes32 beefyMMRLeafHash = hashMMRLeaf(beefyMMRLeaf);
         require(
             lightClientBridge.verifyBeefyMerkleLeaf(
-                beefyMMRLeaf,
+                beefyMMRLeafHash,
                 beefyMMRLeafIndex,
                 beefyMMRLeafCount,
                 peaks,
                 siblings
             ),
-            "Invalid proof"
+            "Channel: Invalid proof"
         );
         verifyMessages(messages, beefyMMRLeaf);
         processMessages(messages);
     }
 
-    // struct BlockHeader {
-    //     parentHash bytes32;
-    //     number uint64;
-    //     stateRoot bytes32;
-    //     extrinsicsRoot bytes32;
-    //     digest bytes;
-    //     messagesRoot bytes32;
-    // }
-    //TODO: verifyMessages should accept all needed proofs
-    function verifyMessages(Message[] memory messages, bytes memory /*beefyMMRLeaf*/)
+    function verifyMessages(Message[] memory messages, BeefyMMRLeaf memory leaf)
         internal
         view
-        returns (bool success)
     {
-
-        // Scale.decodeBeefyMMRLeaf(beefyMMRLeaf)
-        // Scale.decodeBlockHeader(beefyMMRLeaf.BlockHeader)
-        // require(
-        //     blockHeader.BlockNumber <= lightClientBridge.getFinalizedBlockNumber(),
-        //     "block not finalized"
-        // )
-
+        require(
+            leaf.blockNumber <= lightClientBridge.getFinalizedBlockNumber(),
+            "Channel: block not finalized"
+        );
         // Validate that the commitment matches the commitment contents
-        // require(
-        //     validateMessagesMatchCommitment(messages, blockHeader.messageRoot),
-        //     "invalid messages"
-        // );
+        require(
+            validateMessagesMatchRoot(messages, leaf.messagesRoot),
+            "Channel: invalid messages"
+        );
 
         // Require there is enough gas to play all messages
         require(
             gasleft() >= messages.length * MAX_GAS_PER_MESSAGE,
-            "insufficient gas for delivery of all messages"
+            "Channel: insufficient gas for delivery of all messages"
         );
-
-        return true;
     }
 
     function processMessages(Message[] memory messages) internal {
         for (uint256 i = 0; i < messages.length; i++) {
             // Check message nonce is correct and increment nonce for replay protection
-            require(messages[i].nonce == nonce + 1, "invalid nonce");
+            require(messages[i].nonce == nonce + 1, "Channel: invalid nonce");
 
```

Comment:
if some message is too big to be verified, and they cover a range of nonces. Then all the new messages can't be verified because this nonce requirement ?

Reply:
`nonce` must be increase one by one for replay protection.
Darwinia outboud pallet should limit every message size to ensure every message cost less `MAX_GAS_PER_MESSAGE`, and also to ensure total message cost less 1/2 eth block gas limit.

Reply:
Another thing, Can we process all messages and increase `nonce` for batch?